### PR TITLE
Add CMake tests

### PR DESCRIPTION
**Pull request by pdimov**

### CODE REVIEW

**Comment by Lastique** [comment]
I don't really think a build system is something worth testing.

**Comment by pdimov** [comment]
Well, maybe. On the other hand, everything that isn't being tested, breaks earlier or later. Most maintainers even demand tests as part of the CMake pull requests, and rightly so.

**Comment by Lastique** [comment]
I want to minimize maintenance effort spent on CMake. I'm not happy about having two build systems in Boost.Atomic (and other libraries), and so long as we're using Boost.Build as the main build system, I'd like to hear about CMake as little as possible.

**Comment by pdimov** [comment]
Sure, it's your call how to minimize that. Other maintainers (I presume) prefer to, as long as they carry around the CMakeLists.txt file, be sure it works.

But for Atomic specifically, there's another factor in play. There's value in having a model compiled library, one to which we can point people who want to add CMake support to their libraries. For header-only, we can use Assert (f.ex.) as a showcase; for compiled, for reasons already outlined, Atomic is our only choice at the moment.

File: .travis.yml [hunk 1]
```diff
@@ -359,11 +360,50 @@ matrix:
...
+        - cmake -DBOOST_ENABLE_CMAKE=1 -DBoost_VERBOSE=1 -DBOOST_INCLUDE_LIBRARIES=atomic -DBUILD_SHARED_LIBS=ON ..
+        - cmake --build .
+        - ctest --output-on-failure -R boost_atomic
+
+# CMake install test
+    - os: linux
+      env: TEST_CMAKE=1 CMAKE_INSTALL=1
+      script:
+        - mkdir __build__ && cd __build__
+        - cmake -DBOOST_ENABLE_CMAKE=1 -DBoost_VERBOSE=1 -DBOOST_INCLUDE_LIBRARIES="atomic;assert;config;static_assert;type_traits" -DCMAKE_INSTALL_PREFIX=~/.local ..
```
Comment (Lastique) [inline .travis.yml]: I don't like that there is a list of dependencies. Also, install prefix of `~/.local` does not feel right.

File: CMakeLists.txt [hunk 2]
```diff
@@ -37,3 +37,9 @@ if(BOOST_SUPERPROJECT_VERSION)
   boost_install(TARGETS boost_atomic HEADER_DIRECTORY include/)
 
 endif()
+
+if(BUILD_TESTING)
+
+  add_subdirectory(test)
```
Comment (Lastique) [inline CMakeLists.txt]: I don't like that the tests are part of the library CMakeLists.txt.

**Comment by Lastique** [comment]
> Sure, it's your call how to minimize that. Other maintainers (I presume) prefer to, as long as they carry around the CMakeLists.txt file, be sure it works.

Well, I don't really care about CMakeLists.txt, and I don't want to care. If someone uses to build the library, that is their choice. I care about the official build system, which is Boost.Build.

**Comment by Lastique** [comment]
Given that the tests cannot be run anyway because Boost.Thread does not support CMake, and CMakeLists.txt is already tested, I don't think I will be merging this for now.

**Pull request closed without merging**